embedding_bag make_bag_size optimization

ajyu opened the pull request:
Summary:
From James' PR https://github.com/pytorch/pytorch/pull/19715.

embedding_bag microbenchmarks:
Baseline: P123020983
Only refactor make_bag_size, no changing at::zeros to at::empty: P123021393
Both (this diff): P123020983

Inference benchmark on T6_SKL - embedding bag self time only:
bs=40, baseline: .302 ms/iter
bs=40, with diff: .203 ms/iter
bs=1 baseline: .148 ms/iter
bs=1 with diff: .099 ms/iter

The bigger gap comes from fb::embedding_bag_byte_rowwise_offsets, I'm looking into that one too.

Test Plan:
MKL_NUM_THREADS=1 OMP_NUM_THREADS=1 numactl -m 0 -C 3 ./inference_benchmark_nolr_emb.par --pt-scripted-model=traced_model.pt --pt-inputs="batch_size_40/pt_inputs.pth" --iters=3000 --warmup-iters=100

buck run mode/opt //caffe2/benchmarks/operator_benchmark:benchmark_all_other_test -- --tag_filter all --iterations 3000 --operators embeddingbag

Reviewed By: yinghai, qizzzh

Differential Revision: D18714418

yinghai reviewed (approved):
LGTM. Please check the CI failure.

yinghai reviewed (changes requested):
@ajyu The CI test errors are legit, please take a look. 
```
Nov 27 18:44:00 Traceback (most recent call last):
Nov 27 18:44:00   File "/var/lib/jenkins/workspace/test/common_device_type.py", line 196, in instantiated_test
Nov 27 18:44:00     result = test(self, device_arg, dtype)
Nov 27 18:44:00   File "test_nn.py", line 9821, in test_embedding_bag_device
Nov 27 18:44:00     self._test_EmbeddingBag(device, 'sum', False, dtype)
Nov 27 18:44:00   File "test_nn.py", line 9768, in _test_EmbeddingBag
Nov 27 18:44:00     output.backward(grad_output_with_empty)
Nov 27 18:44:00   File "/opt/conda/lib/python3.6/site-packages/torch/tensor.py", line 195, in backward
Nov 27 18:44:00     torch.autograd.backward(self, gradient, retain_graph, create_graph)
Nov 27 18:44:00   File "/opt/conda/lib/python3.6/site-packages/torch/autograd/__init__.py", line 99, in backward
Nov 27 18:44:00     allow_unreachable=True)  # allow_unreachable flag
Nov 27 18:44:00 RuntimeError: Expected a Tensor of type Variable but found an undefined Tensor for argument #4 'bag_size'
```

ajyu commented:
Got rid of unintended commit.

ajyu commented:
Tests fail because embedding_bag backward requires the bag_size tensor in its signature https://fburl.com/diffusion/w24il4a4.

ajyu commented:
In MODE_SUM, initialize bag_size only if we need gradients.

ajyu committed:
embedding_bag make_bag_size optimization (#30495)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/30495

From James' PR https://github.com/pytorch/pytorch/pull/19715.

embedding_bag microbenchmarks:
Baseline: P123020983
Only refactor make_bag_size, no changing at::zeros to at::empty: P123021393
Both (this diff): P123020983

Inference benchmark on T6_SKL - _embedding_bag self time only:
bs=40, baseline: .302 ms/iter
bs=40, with diff: .203 ms/iter
bs=1 baseline: .148 ms/iter
bs=1 with diff: .099 ms/iter

The bigger gap comes from fb::embedding_bag_byte_rowwise_offsets, I'm looking into that one too.

Test Plan:
MKL_NUM_THREADS=1 OMP_NUM_THREADS=1 numactl -m 0 -C 3 ./inference_benchmark_nolr_emb.par --pt-scripted-model=traced_model.pt --pt-inputs="batch_size_40/pt_inputs.pth" --iters=3000 --warmup-iters=100

buck run mode/opt //caffe2/benchmarks/operator_benchmark:benchmark_all_other_test -- --tag_filter all --iterations 3000 --operators embeddingbag

Reviewed By: yinghai, qizzzh

Differential Revision: D18714418

fbshipit-source-id: c9c9fa6b85653d8d500a318dcaad985b7425afc1

ajyu commented:
Close this PR to try the export again.
